### PR TITLE
[fix] 'Expected End Date' in task should not be later than 'Expected End Date' of project

### PR DESCRIPTION
**Pull request by Anurag810**
Please read the pull request checklist to make sure your changes are merged: https://github.com/frappe/erpnext/wiki/Pull-Request-Checklist

### CODE REVIEW

File: erpnext/projects/doctype/task/task.py [hunk 1]
```diff
@@ -43,6 +43,10 @@ def validate_dates(self):
 		if self.act_start_date and self.act_end_date and getdate(self.act_start_date) > getdate(self.act_end_date):
 			frappe.throw(_("'Actual Start Date' can not be greater than 'Actual End Date'"))
 
+		doc = frappe.get_doc("Project",self.project)
+		if  self.exp_end_date and doc.expected_end_date and getdate(self.exp_end_date) > getdate(doc.expected_end_date) :
+			frappe.throw(_("Expected end date cannot be after Project: "+doc.name+" Expected end date"))
```
Comment (nabinhait) [inline erpnext/projects/doctype/task/task.py]: This is not the pattern we use for message, will create problem for translation.

File: erpnext/projects/doctype/task/task.py [hunk 2]
```diff
@@ -43,6 +43,10 @@ def validate_dates(self):
 		if self.act_start_date and self.act_end_date and getdate(self.act_start_date) > getdate(self.act_end_date):
 			frappe.throw(_("'Actual Start Date' can not be greater than 'Actual End Date'"))
 
+		doc = frappe.get_doc("Project",self.project)
```
Comment (nabinhait) [inline erpnext/projects/doctype/task/task.py]: space after comma.

**Review by nabinhait — CHANGES_REQUESTED** [review]
test case missing

**Comment by nabinhait** [comment]
codacy failing

File: erpnext/projects/doctype/task/task.py (hunk 1 above)
Comment (Anurag810) [reply erpnext/projects/doctype/task/task.py]: changed

File: erpnext/projects/doctype/task/task.py (hunk 2 above)
Comment (Anurag810) [reply erpnext/projects/doctype/task/task.py]: done

File: erpnext/projects/doctype/task/task.py [hunk 3]
```diff
@@ -12,6 +12,7 @@
 
 
 class CircularReferenceError(frappe.ValidationError): pass
+class EndDateConnotGreaterThanProjecteEndDateError(frappe.ValidationError): pass
```
Comment (Alchez) [inline erpnext/projects/doctype/task/task.py]: A couple of typos here:

- `Connot` -> `Cannot`
- `EndDateConnotGreater` -> `EndDateCannotBeGreater`
- `ProjecteEndDate` -> `ProjectEndDate`

File: erpnext/projects/doctype/task/task.py [hunk 4]
```diff
@@ -43,6 +44,11 @@ def validate_dates(self):
 		if self.act_start_date and self.act_end_date and getdate(self.act_start_date) > getdate(self.act_end_date):
 			frappe.throw(_("'Actual Start Date' can not be greater than 'Actual End Date'"))
 
+		doc = frappe.get_doc("Project", self.project)
+		if  self.exp_end_date and doc.expected_end_date and getdate(self.exp_end_date) > getdate(doc.expected_end_date) :
+			frappe.msgprint(_("Expected end date cannot be after Project: <b>'{0}'</b> Expected end date").format(doc.name))
```
Comment (Alchez) [inline erpnext/projects/doctype/task/task.py]: Do a `frappe.throw` instead, with the exception.

**Review by Alchez — CHANGES_REQUESTED** [review]

File: erpnext/projects/doctype/task/task.py (hunk 4 above)
Comment (Anurag810) [reply erpnext/projects/doctype/task/task.py]: Done

File: erpnext/projects/doctype/task/task.py (hunk 3 above)
Comment (Anurag810) [reply erpnext/projects/doctype/task/task.py]: Done

File: erpnext/projects/doctype/task/task.py [hunk 5]
```diff
@@ -43,6 +44,12 @@ def validate_dates(self):
 		if self.act_start_date and self.act_end_date and getdate(self.act_start_date) > getdate(self.act_end_date):
 			frappe.throw(_("'Actual Start Date' can not be greater than 'Actual End Date'"))
 
+		if(self.project):
+			if frappe.db.exists("Project", self.project):
+				doc = frappe.get_doc("Project", self.project)
```
Comment (nabinhait) [inline erpnext/projects/doctype/task/task.py]: Avoid using `get_doc` for getting only a field's value, use `frappe.db.get_value` instead.

**Review by nabinhait — CHANGES_REQUESTED** [review]

File: erpnext/projects/doctype/task/task.py (hunk 5 above)
Comment (Anurag810) [reply erpnext/projects/doctype/task/task.py]: Done

**Pull request merged**